### PR TITLE
assistant-evals: an MCP eval driven by a real Claude Code subprocess

### PR DESCRIPTION
**Pull request by dmaretskyi**
The CLI's `mcp/agent-e2e.test.ts` shape — a real `claude` subprocess against this repo's MCP surface, every claim read back out of the database — as an eval, with the server hosted **inside the eval process** instead of as `dx mcp serve`.

## What's here

- **`assistant-evals/src/mcp-host.ts`** — serves the projected surface (`queryOperations` / `invokeOperation` / `loadSkill`) over Streamable HTTP on a loopback port. Every tool answers through `@dxos/mcp-server`'s own dispatch against the harness's registry and operation invoker, so the eval measures the deployed surface — governance by skill, space resolution, input validation, ref qualification — not a reimplementation. The transport is the MCP SDK's; the harness supplies a context carrying the space's database, because a reference argument arrives as a wire envelope and a `Ref` decoded without one has no resolver.
- **`assistant-evals/src/claude-harness.ts`** — `runClaudeEval({ skills, plugins, types, seed }, body)`: boots a Composer harness, starts that server against it, and spawns `claude` wired to it with only the server's three verbs allowed (no Bash, no file tools, so a prompt the surface cannot satisfy fails). `body` gets `send(prompt)` for one agent turn, `query(effect)` for a read in the harness's own runtime outside the agent, and `score(scorers)` to grade the still-open space.
- **`assistant-evals/src/evals/mcp-server.eval.ts`** — six dimensions on the `Scorer` API: the scaffold before any tokens are spent, a read-only turn that must change nothing, one targeted write with the untouched task as the control, a follow-up turn that must not roll the first one back, and the ledger's length (an agent that cannot find a task tends to create one and report success). Five are staged facts a turn established and the final space cannot recover; `ledger-intact` is end state and queries the database itself.
- **`test-utils/src/claude-agent.ts`** — the Claude Code driver, moved out of the CLI's testing directory now that both its end-to-end test and this eval drive it. A turn is single-flight and a timeout is terminal, so a late `result` cannot land on the next turn. It lives in a library package because moon's layering forbids `cli` (application) depending on `assistant-evals` (automation).
- **`assistant-evals/src/mcp-host.test.ts`** — deterministic and offline: drives the server with the same MCP client `AiSession` uses, asserting the three tools list and that `invokeOperation` reaches the invoker with the right key, input and space. The package spells out its own `test` task, since the `ts-test` tag's `--project=node` filter matches nothing against the flat vitest config evalite needs.

No CLI binary, no bootstrapped profile, one database: each stage reads the space the agent just wrote to.

## Test plan

- `npx evalite run src/evals/mcp-server.eval.ts` — **100%** against `sonnet`, all six scorers. Needs `DX_ANTHROPIC_API_KEY` and a `claude` binary; `DX_EVAL_CLAUDE_MODEL` overrides the model.
- `moon run assistant-evals:test` — the host test passes through moon, no model.
- `moon run assistant-evals:lint test-utils:lint cli:lint`, `npx tsc --noEmit` — clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Bce52MMJNMucFANFASmpLQ

---
🚀 Composer preview: https://pr-13059-composer-dev.dxos.workers.dev

### CODE REVIEW

File: packages/core/compute/assistant-evals/src/mcp-host.test.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+/**
+ * The in-process host an eval's agent dials, exercised by the same MCP client `AiSession` uses.
+ *
+ * Deterministic and offline — no model, no harness — because what can break here is the transport:
+ * effect's RPC layer answers a single request with a one-element batch, and a client that receives
+ * the array lists no tools at all (see `McpServer.normalizeResponse`).
+ */
+describe('startMcpHost', () => {
+  test('serves the projected surface to an MCP client and dispatches to the invoker', async ({ expect }) => {
```
[the remark this answers is not in the document]
Comment (dmaretskyi) [reply packages/core/compute/assistant-evals/src/mcp-host.test.ts:75]: Stale as of the current head: the eval no longer goes through the assistant session at all. The agent is a real `claude` subprocess talking to the in-process server, so `AiSession` / `McpToolkit.make` are not on the path any more, and the eval scores 100% against sonnet.

The underlying finding still stands as its own bug, though, and this PR is not where it gets fixed: connecting any MCP server to an assistant session kills the run inside effect's `Schema` (`Cannot read properties of undefined (reading 'encoding')`, from `ProcessHandle`'s input encode). Bisected before the rework — the same run passes when the server answers nothing, fails with zero skills served, and failed identically under effect's own `McpServer.layerHttp` transport. That points at `McpToolkit.make` building every tool parameter from the shared `Schema.Unknown`, and it should hit the production `browser` skill too. Worth a separate PR with an assistant-session regression test.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/core/compute/assistant-evals/src/claude-harness.ts [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        model: options.model ?? DEFAULT_MODEL,
+        allowedTools: options.allowedTools ?? [tool('queryOperations'), tool('invokeOperation'), tool('loadSkill')],
+        timeout: options.turnTimeout ?? DEFAULT_TURN_TIMEOUT,
+      });
+      const claudeAgent = agent;
+
+      return await body({ spaceId, query, send: (prompt) => claudeAgent.send(prompt) });
+    } finally {
+      await Effect.runPromise(Scope.close(scope, Exit.void));
+    }
```
[the remark this answers is not in the document]
Comment (dmaretskyi) [reply packages/core/compute/assistant-evals/src/claude-harness.ts]: Fixed in `387e0e30` — the scope is now opened immediately before the `try` that closes it, so a host that dies after binding the listener still runs its finalizer. Kept it non-optional rather than the proposed `let scope … if (scope)`: `Scope.make()` itself can't fail, so the only gap was the host startup, and covering that keeps the type honest.

Two more from the same review, also in `387e0e30`:

- `send` is single-flight now (a second call while a turn is in flight is rejected instead of taking over the first one's callbacks), and a timeout goes through `#fail`, so a late `result` from the timed-out turn cannot land on the next one. Good catch — the conversation is one stream and the old code assumed a disciplined caller.
- JSDoc on `start`/`close`, and the `runClaudeEval` block moved above the function rather than above the middleware helper that had drifted between them.

Not doing the fourth (schema-decoding every stream event before storing it): the payloads come from the `claude` CLI this repo spawns itself, not from an untrusted peer, and drift surfaces immediately as a turn that fails its scorer rather than as a silent wrong answer — `Turn` reads three fields off events the CLI documents. A tagged-schema decode layer over that is a real change to a test driver's contract and belongs in its own PR if the CLI's stream ever proves unstable.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/core/compute/assistant-evals/src/claude-harness.ts [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        // server of its own — the difference between measuring this surface and measuring a CLI.
+        mcpServers: { [SERVER]: { type: 'http', url } },
+        apiKey: API_KEY,
+        model: options.model ?? DEFAULT_MODEL,
+        allowedTools: options.allowedTools ?? [tool('queryOperations'), tool('invokeOperation'), tool('loadSkill')],
+        timeout: options.turnTimeout ?? DEFAULT_TURN_TIMEOUT,
+      });
+      const claudeAgent = agent;
+
+      const started = Date.now();
```
[the remark this answers is not in the document]
Comment (dmaretskyi) [reply packages/core/compute/assistant-evals/src/claude-harness.ts:202]: Right, and the comment above it claimed otherwise — `started` was taken before `body`, so `durationMillis` covered the queries between turns and the scaffolding too. Fixed as you suggest: the timer starts inside `send` and each turn's elapsed time is added, so `Scorer.runAll` gets the sum of the turns and nothing else.

Also removing `--passWithNoTests` from the package's test task (your other comment): this package has exactly one test file, so discovering none is a config regression, not a shard with nothing to run. `moon run assistant-evals:test` → `Test Files 1 passed (1)` without the flag.

Both are committed locally as `b176bd6f` but **not yet pushed**: the PR is in the merge queue and the branch is locked (`protected branch hook declined`). They will land in a follow-up PR straight after this one merges, rather than dequeuing a green PR for two non-blocking fixes. Neither affects the eval's result — no scorer here reads `durationMillis`, and the test task passes either way.

---
_Generated by [Claude Code](https://claude.ai/code)_
